VocaDbModel/Service/QueryableExtenders/VenueQueryableExtender.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Linq;
+using VocaDb.Model.Domain.Globalization;
+using VocaDb.Model.Domain.Venues;
+using VocaDb.Model.Service.Search;
+
+namespace VocaDb.Model.Service.QueryableExtenders {
+
+	public static class VenueQueryableExtender {
+
+		private static double GetEarthRadius(DistanceUnit unit) => unit switch
+		{
+			DistanceUnit.Kilometers => 6371,
+			DistanceUnit.Miles => 3960,
+			_ => throw new ArgumentException()
+		};
+
+		public static IQueryable<Venue> OrderBy(this IQueryable<Venue> query, VenueSortRule sortRule, ContentLanguagePreference languagePreference, GeoPointQueryParams coordinates, DistanceUnit distanceUnit) {
+
+			switch (sortRule) {
+				case VenueSortRule.Name:
+					return query.OrderByName(languagePreference);
+				case VenueSortRule.Distance:
+					return query.OrderByDistance(coordinates, distanceUnit);
+			}
+
+			return query;
+
+		}
+
+		public static IQueryable<Venue> OrderByDistance(this IQueryable<Venue> query, GeoPointQueryParams coordinates, DistanceUnit distanceUnit) {
+
+			if (!coordinates.HasValue)
+				return query;
+
+			var earthRadius = GetEarthRadius(distanceUnit);
+
+			// HACK: calculate the Haversine distance.
+			// TODO: do something.
+			return query.OrderBy(v => earthRadius * 2 * Math.Asin(Math.Sqrt(
+				Math.Pow(Math.Sin((coordinates.Latitude.Value - Math.Abs(v.Coordinates.Latitude.Value)) * Math.PI / 180 / 2), 2)
+				+ Math.Cos(coordinates.Latitude.Value * Math.PI / 180)
+				* Math.Cos(Math.Abs(v.Coordinates.Latitude.Value) * Math.PI / 180)
+				* Math.Pow(Math.Sin((coordinates.Longitude.Value - v.Coordinates.Longitude.Value) * Math.PI / 180 / 2), 2))));

```

Comment:
We have to do something here.

Reply:
What does the current code do?

Reply:
In order to calculate the distance between two points on a sphere, I applied the [haversine formula](https://en.wikipedia.org/wiki/Haversine_formula) in a very ugly way because NHibernate won't understand property calls as you said. I actually wanted to use geography data types and NHibernate.Spatial as I said before, but they don't work well. Please see: https://build-failed.blogspot.com/2012/03/nhibernate-spatial-part-4-querying.html

Reply:
If it works then it's fine to keep it like this I think? It's not that ugly I think. You could mention in the comment what you just said, and you can remove the "HACK: do something" comment, I think.